### PR TITLE
Director enhancements

### PR DESCRIPTION
**Pull request by CannonLock**

### CODE REVIEW

File: web_ui/frontend/app/director/components/DirectorCardList.tsx [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+                    </Grid>
+                </Grid>
+            </Box>
+            <CardList data={filteredData} Card={DirectorCard} cardProps={cardProps}/>
+        </Box>
+    )
+}
+
+const isPelicanServer = (server?: Server) => {
+    return server?.webUrl != "";
```
Comment (haoming29) [inline web_ui/frontend/app/director/components/DirectorCardList.tsx:68]: With the merged change to the director response, now there's a dedicated field `FromTopology` for this.

File: web_ui/frontend/app/director/components/DirectorCardList.tsx [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+const serverHasError = (server?: Server) => {
+    return server?.status == "Error";
```
Comment (haoming29) [inline web_ui/frontend/app/director/components/DirectorCardList.tsx:72]:
```suggestion
    return server?.status === "Error";
```

File: web_ui/frontend/app/director/components/DirectorCardList.tsx [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    const [pelicanServer, setPelicanServer] = useState<boolean | undefined>(undefined)
+    const [serverError, setServerError] = useState<boolean | undefined>(undefined)
+    const [serverDowntime, setServerDowntime] = useState<boolean | undefined>(undefined)
+
+    const searchedData = useFuse<Partial<DirectorCardProps>>(data, search)
+
+    const filteredData = useMemo(() => {
+        let filteredData = structuredClone(searchedData)
+        if (pelicanServer != undefined) {
+            filteredData = filteredData.filter((d) => isPelicanServer(d?.server) == pelicanServer)
```
Comment (haoming29) [inline web_ui/frontend/app/director/components/DirectorCardList.tsx:30]:
```suggestion
            filteredData = filteredData.filter((d) => isPelicanServer(d?.server) && pelicanServer)
```
why not using `&&` for the three filters?

**Comment by haoming29** [comment]
@CannonLock Could you provide the issue(s) this PR can close?

File: web_ui/frontend/app/director/components/DirectorCard.tsx
Comment (haoming29) [inline web_ui/frontend/app/director/components/DirectorCard.tsx:1]: Do you want to include tooltip for the switch in this PR as well? https://github.com/PelicanPlatform/pelican/issues/1072

Comment (haoming29) [inline web_ui/frontend/app/director/components/DirectorCard.tsx:1]: Also, reading through the list in the original ticket: https://github.com/PelicanPlatform/pelican/issues/1249 Brian suggests adding server status as part of the DirectorCard, i.e. showing the status at the top-level as well.

File: web_ui/frontend/app/director/page.tsx [hunk 4]
```diff
@@ -63,7 +48,7 @@ export default function Home() {
                 <Grid item xs={12} lg={8} xl={6}>
                     <Typography variant={"h4"} pb={2}>Origins</Typography>
                     {originData ?
```
Comment (haoming29) [inline web_ui/frontend/app/director/page.tsx:50]: If the filter returns 0 entries, I got the skeleton animation:
Screenshot 2024-06-11 at 10 05 21 AM
While I think the user is expecting an empty list with text suggesting no result found

File: web_ui/frontend/components/BooleanToggleButton.tsx [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    value?: boolean;
+    onChange: (value?: boolean) => void;
+}
+
+export const BooleanToggleButton = ({label, value, onChange}: BooleanToggleButtonProps) => {
+    return (
+        <ButtonGroup
+            aria-label={label}
+        >
+            <Button disabled size={"small"}>
```
Comment (haoming29) [inline web_ui/frontend/components/BooleanToggleButton.tsx:16]: Can we pass a brief description to the label as tooltips? At least `Pelican Server` sounds a bit confused to users not knowing the difference

File: web_ui/frontend/app/director/components/DirectorDropdown.tsx [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    transition: boolean;
+}
+
+export const DirectorDropdown = ({server, transition} : DirectorDropdownProps) => {
+    return (
+        <Dropdown transition={transition} flexDirection={"column"}>
+            <InformationSpan name={"Type"} value={server.type} />
+            <InformationSpan name={"Status"} value={server.status} />
+            <InformationSpan name={"URL"} value={server.url} />
+            <Typography variant={"body2"} sx={{fontWeight: 500,  display: "inline", mr: 2, my: .5}}>Namespace Prefixes</Typography>
```
Comment (haoming29) [inline web_ui/frontend/app/director/components/DirectorDropdown.tsx:18]: We probably also want to include
* Server geolocation (`longitude` and `latitude`)
* Server capability (`Caps` field, similar to the origin data exports)

File: web_ui/frontend/index.d.ts [hunk 7]
```diff
@@ -6,3 +8,41 @@ export interface User {
...
+    "webUrl": string;
+    "type": "Origin" | "Cache";
+    "latitude": number;
+    "longitude": number;
+    "enableWrite": boolean;
+    "enableFallbackRead": boolean;
+    "filtered": boolean;
+    "filteredType": string;
+    "status": string;
+    "namespacePrefixes": string[];
```
Comment (haoming29) [inline web_ui/frontend/index.d.ts:26]: Please refer to the comments above for changes to these fields.

File: web_ui/frontend/components/CardList.tsx [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                    } as T
+
+                    return <Box pb={1} key={JSON.stringify(props)}>
+                        <Card {...props} />
+                    </Box>
+                })}
+                {!data || count == 0 && <Skeleton variant="rectangular" height={60}></Skeleton>}
+            </Box>
+            <Box display={"flex"} justifyContent={"center"} flexDirection={"column"} alignItems={"center"}>
+                { count > 1 && <Pagination count={count} page={page} onChange={(e, p) => setPage(p)} /> }
```
Comment (haoming29) [inline web_ui/frontend/components/CardList.tsx:40]: This count seems to be off. Without filtering anything, I got 5 entries in one page, but the number shows 8.

Screenshot 2024-06-11 at 10 05 26 AM
I thought the real "count" here is the total number of servers in the table, as we don't really need a count for servers per page.

Looking at MUI docs, `count` is referring to "total number of pages": https://mui.com/material-ui/api/pagination/

**Review by haoming29 — CHANGES_REQUESTED** [review]

File: web_ui/frontend/app/director/components/DirectorCard.tsx
Comment (CannonLock) [reply web_ui/frontend/app/director/components/DirectorCard.tsx:1]: @haoming29 The card shows up orange if it is in error state.

File: web_ui/frontend/app/director/components/DirectorCardList.tsx (hunk 3 above)
Comment (CannonLock) [reply web_ui/frontend/app/director/components/DirectorCardList.tsx:30]: The BooleanToggleButton's value is undefined | true | false based on what value the user wants the associated value to be in the data. 

If I used && then if the user wanted to see only a falsy field they wouldn't be able too.

**Comment by CannonLock** [comment]
Screenshot 2024-06-13 at 12 19 26 PM (2)

**Review by haoming29 — APPROVED** [review]
Looks really nice to me! Some comments that are non-blocking so I'll conditionally approve:
* For an errored server, when I hover on the server bar, the orange color goes away:
Screenshot 2024-06-14 at 9 12 09 AM
Screenshot 2024-06-14 at 9 12 05 AM

* Can we move the capabilities above the namespace prefixes? I can imagine a large spanning tree of namespace prefix can block the view on capabilities
* When I navigate to non-first page, and click on the filter, and if the result is less than the page number I was on, I will get nothing:
Screenshot 2024-06-14 at 9 13 07 AM
I think a simple fix can be just reset the page number to 1 when the user toggles the filter

**Comment by CannonLock** [comment]
@haoming29 I will fix these up, thanks for catching.

**Review by haoming29 — APPROVED** [review]
LGTM!

**Pull request merged**